crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,534 @@
+pub struct IpcProviderConfig {
+    pub(crate) request_timeout_ms: u64,
+    pub(crate) ipc_path: PathBuf,
+    pub(crate) mempool_server_url: String,
+}
+
+impl Default for IpcProviderConfig {
+    fn default() -> Self {
+        Self {
+            request_timeout_ms: DEFAULT_IPC_REQUEST_TIMEOUT_MS,
+            mempool_server_url: String::new(),
+            ipc_path: PathBuf::new(),
+        }
+    }
+}
+
+impl StateProviderFactory for IpcStateProviderFactory {
+    /// Gets state for the latest block
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            BlockNumberOrTag::Latest.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block number
+    // We are not caching state provider by block number to avoid any issues with reorgs
+    // The calls to  history_by_block_number are rare and we shouldn't be loosing perf by not
+    // leveraging caching here
+    fn history_by_block_number(&self, block: BlockNumber) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block hash
+    fn history_by_block_hash(&self, block: BlockHash) -> ProviderResult<StateProviderBox> {
+        if let Some(state) = self.state_provider_by_hash.get(&block) {
+            return Ok(Box::new(state));
+        }
+
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        self.state_provider_by_hash.insert(block, *state.clone());
+        Ok(state)
+    }
+
+    /// Gets block header given block hash
+    fn header(&self, block_hash: &BlockHash) -> ProviderResult<Option<Header>> {
+        let header = rpc_call::<
+            _,
+            Option<<alloy_network::Ethereum as alloy_network::Network>::BlockResponse>,
+        >(
+            &self.ipc_provider,
+            "eth_getBlockByHash",
+            (block_hash, false),
+        )?
+        .map(|b| b.header.inner);
+
+        Ok(header)
+    }
+
+    /// Gets block hash given block number
+    fn block_hash(&self, number: BlockNumber) -> ProviderResult<Option<B256>> {
+        let block_hash = rpc_call::<_, Option<B256>>(
+            &self.ipc_provider,
+            "rbuilder_getBlockHash",
+            (BlockNumberOrTag::Number(number),),
+        )?;
+
+        Ok(block_hash)
+    }
+
+    /// Gets block number of latest known block
+    fn best_block_number(&self) -> ProviderResult<BlockNumber> {
+        self.last_block_number()
+    }
+
+    /// Gets block header given block hash
+    fn header_by_number(&self, num: u64) -> ProviderResult<Option<Header>> {
+        let block = rpc_call::<
+            _,
```

Comment:
I'm in favor of using concrete type here. This will help in case we will do any refactoring and accidentally change parameters. This way we would be able to find problem instantly.

Reply:
Sorry, I completely missed the comment 🤦‍♂️
I disagree; the type here is inferred from the function arguments, i.e., `(num, false). ' If we mess it up, rust will yell at us. 
That being said, I don't feel strongly about this if @ZanCorDX  agrees with you, I'll refactor :)

Reply:
hm, will it yell if we change params to (num, false, false) for example? If yes - all good
But i assume it will send them to rpc and get error because there is extra false param

Reply:
Correct, it won't :)
So the idea is that if we have written this out in 2 places (in the type and in the args) there is smaller chance of us making mistake in both places when refactoring?

Reply:
@SozinM done :)
Sorry it took me so long!

https://github.com/flashbots/rbuilder/pull/489/commits/fbe50877205181b7f961afd17379c043b6c339d1